## Functions- This is reusable block of code . 
* This is block of code where some instruction are stored that is used to perform some task.

# Inbuilt function:-
* The function that is already defined by the developer is known as inbuilt function we can use it but we cannot define it .
* Types Inbuilt function:-

1. Utility function - the function which is common for each and every datatype is known as utility  function eg-(len(),dir(),type(),bool(),id()). 

2. On string - the function that we can only apply on string datatype is known as inbuilt function on string.
* lower():- it is used to convert upper case to lower 
* syntax - var_n.lower()
* Upper():- it is used to convert into lower case to upper
* syntax :- var_n.upper()
* islower():- it is used to check lower case value
* syntax :- var_n.islower()
* isupper():- it is used to check upper case value
* syntax:-var_n.isupper()
* swapcase():- it is used to swap the cases
* syntax:- var_n.swapcase()
* title():- first character of each word into upper case
* syntax :- var_n.title()
* isdigit():- used to check the ch is digit or not
* syntax :- var_n.isdigit()

3. 
4. 
5. 
6. 

# user define function
* Function that is define by the user as per requirement is known as user define function

```python
# syntax- 
def function_name(args):
    {
        code block 
    }

function_name() 
```

##### Types of user defined function:
```python 

return :- It is a keyword that is used to stop the function and return some values to the function.

1. without arg and return values
# syntax:-
def f_name():
    (code)
f_name()

2. with arg and without return values
(user define function having argument but not having return values)
# syntax:-
def f_name(args):
    (code)
f_name(values)

3. without args and with return values 
(function does not have argument but haves return values )
# syntax:-
def f_name():
    (code)
    return values
var_name=f_name() or print(f_name())

4. with args and with return values
(functions have arguments and also has return values )
#syntax:-
def f_name(args):
    (code)
    return values
var_name=f_name(values)
example:-
def sum(n):
    sum=0 
    for i in range(n):
        sum=sum+i
    return sum
print(sum())
a=sum()
```


```python
                           memory allocation
            main space              |         method space
                                    |
                 add                |         0x11
                 0x11               |    |a=input()|
                                    |    |b=input()|
                                    |    |a+b      |
                                    | 
```


121.Wap to check whether the number is strong or not. 

In [77]:
def strong(s):
    def fac(digit):
        fact=1
        for i in range(1,digit+1):
            fact=fact*i
        return fact
    def sum(s):
        sum=0
        while s>0:
            digit=s%10
            sum=sum+fac(digit)
            s=s//10
        return sum
    if s == sum(s):
        print("strong number")
    else:
        print("not a strong number")

s=int(input("enter the number: "))
strong(3)

    

not a strong number


122.Wap to print the strong numbers between the given range.

In [104]:
def n_strong():
    def strong(s):
        def fac(digit):
            fact=1
            for i in range(1,digit+1):
                fact=fact*i
            return fact
        def sum(s):
            sum=0
            while s>0:
                digit=s%10
                sum=sum+fac(digit)
                s=s//10
            return sum
        if s == sum(s):
            print(f"strong number {s}")
            # return s
        # else:
        #     print("not a strong number")
    a=int(input("enter the range: "))
    for i in range(1,a+1): 
        strong(i)
n_strong()


strong number 1
strong number 2


123.Wap to find the nth strong number.

In [123]:
def nth_strong(x,a):    
    def strong(s):
        def fac(digit):
            fact=1
            for i in range(1,digit+1):
                fact=fact*i
            return fact
        def sum(s):
            sum=0
            while s>0:
                digit=s%10
                sum=sum+fac(digit)
                s=s//10
            return sum
        if s == sum(s):
            return s 
        # else:
        #     print("not a strong number")
    last_f=[]
    for i in range(1,x+1): 
        s=strong(i)
        if s is not None:
            last_f.append(s)
    if a<len(last_f):
        print(last_f[a])
    else:
        print("count of strong number ",len(last_f)-1)
        
nth_strong(1000,0)
            

1


124.Wap to check whether the number is arm strong or not. 

In [122]:
def armstrong(n):    
    def count(n):
        count=0
        while n>0:
            count=count+1
            n=n//10
        return count
    def sum(n):
        sum=0
        power=count(n)
        while n>0:
            digit=n%10
            sum=sum+(digit**power)
            n=n//10
        return sum
    if sum(n) == n:
        print("armstrong")
    else:
        print("not armstrong ")
armstrong(153)
    

armstrong


125.Wap to print the arm strong numbers between the given range.

In [74]:
def n_armstrong(n):   
    def armstrong(n):    
        def count(n):
            count=0
            while n>0:
                count=count+1
                n=n//10
            return count
        def sum(n):
            sum=0
            power=count(n)
            while n>0:
                digit=n%10
                sum=sum+(digit**power)
                n=n//10
            return sum
        if sum(n) == n:
            print(f"armstrong,{n}")
        else:
            print(f"not armstrong ,{n} ")
    for i in range(1,n+1):
        armstrong(i)
n_armstrong(1)

    

armstrong,1


126.Wap to find the nth arm strong number. 

In [27]:
def nth_armstrong(n):   
    def armstrong(n):    
        def count(n):
            count=0
            while n>0:
                count=count+1
                n=n//10
            return count
        def sum(n):
            sum=0
            power=count(n)
            while n>0:
                digit=n%10
                sum=sum+(digit**power)
                n=n//10
            return sum
        if sum(n) == n:
            return n
        # else:
        #     print(f"not armstrong ,{n} ")
    last_n=None
    for i in range(1,n+1):
        s=armstrong(i)
        if s is not None:
            last_n=s
    return last_n
nth_armstrong(153)

153

127.Wap to check whether the number is perfect or not.

In [ ]:
def perfect_number():
    def sum(n):
        sum=0
        for i in range(1,n):
            if n % i == 0:
                sum=sum+i
        return sum
    n=6
    if sum(n) == n:
        print("perfect number")
    else:
        print("not a perfect number")
perfect_number()        
    

perfect number


128.wap to print the perfect numbers between the given range. 

In [51]:
def n_perfect_number(n):
    def perfect_number(n):
        def sum(n):
            sum=0
            for i in range(1,n):
                if n%i == 0:
                    sum=sum+i
            return sum
        if sum(n) == n:
            print(f"perfect number ,{n}")
        # else:
        #     print("not a perfect number")
    last_n=None
    for i in range(1,n+1):
        s=perfect_number(i)
        if s is not None:
            last_n=s
    return last_n
n_perfect_number(100)

perfect number ,6
perfect number ,28


129.Wap to find the nth perfect number.

In [55]:
def nth_perfect_number(n):
    def perfect_number(n):
        def sum(n):
            sum=0
            for i in range(1,n):
                if n%i == 0:
                    sum=sum+i
            return sum
        if sum(n) == n:
            return n 
        # else:
        #     print("not a perfect number")
    last_n=None
    for i in range(1,n+1):
            s=perfect_number(i)
            if s is not None:
                last_n=s
    return last_n            
nth_perfect_number(100)

28

130.Wap to check whether the number is prime or not.

In [64]:
def prime(n): 
    count=0
    for i in range(1,n+1):
        if n%i==0:
            count=count+1     
    if count<=2:
        print("prime")
    else:
        print("not prime")
prime(11)
    
        
    

prime


131.wap to print the prime numbers between the given range.

In [71]:
def n_prime(n):
    def prime(n): 
        count=0
        for i in range(1,n+1):
            if n%i==0:
                count=count+1     
        if count<=2:
            print(f"prime ,{n}")
        # else:
        #     print("not prime")
    for i in range(1,n+1):
        prime(i)
n_prime(20)

    

prime ,1
prime ,2
prime ,3
prime ,5
prime ,7
prime ,11
prime ,13
prime ,17
prime ,19


132.Wap to find the nth prime number.

In [113]:
def nth_prime(n,a):
    def prime(n): 
        count=0
        for i in range(1,n+1):
            if n%i==0:
                count=count+1     
        if count<=2:
            return n
        # else:
        #     print("not prime")
    last_n=[]
    for i in range(1,n+1):
        s=prime(i)
        if s is not None:
            last_n.append(s)
    if a<len(last_n):
        return last_n[a]
    else:
        return last_n
            
nth_prime(1000,999)

[1,
 2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97,
 101,
 103,
 107,
 109,
 113,
 127,
 131,
 137,
 139,
 149,
 151,
 157,
 163,
 167,
 173,
 179,
 181,
 191,
 193,
 197,
 199,
 211,
 223,
 227,
 229,
 233,
 239,
 241,
 251,
 257,
 263,
 269,
 271,
 277,
 281,
 283,
 293,
 307,
 311,
 313,
 317,
 331,
 337,
 347,
 349,
 353,
 359,
 367,
 373,
 379,
 383,
 389,
 397,
 401,
 409,
 419,
 421,
 431,
 433,
 439,
 443,
 449,
 457,
 461,
 463,
 467,
 479,
 487,
 491,
 499,
 503,
 509,
 521,
 523,
 541,
 547,
 557,
 563,
 569,
 571,
 577,
 587,
 593,
 599,
 601,
 607,
 613,
 617,
 619,
 631,
 641,
 643,
 647,
 653,
 659,
 661,
 673,
 677,
 683,
 691,
 701,
 709,
 719,
 727,
 733,
 739,
 743,
 751,
 757,
 761,
 769,
 773,
 787,
 797,
 809,
 811,
 821,
 823,
 827,
 829,
 839,
 853,
 857,
 859,
 863,
 877,
 881,
 883,
 887,
 907,
 911,
 919,
 929,
 937,
 941,
 947,
 953,
 967,
 971,
 977,
 983,
 991,
 997]

133.Wap to check whether the number is Fibonacci number or not. 

In [80]:
def f(n,a=0,b=1):
    if n<0:
        return False
    if n == a or n == b:
        return True
    if b>n:
        return False
    return f(n,b,a+b)
f(11)


False

134.wap to print the Fibonacci numbers between the given range. 

In [98]:
def f(n):
    if n<=1:
        return n
    a=f(n-1)
    b=f(n-2)
    return a+b
f(8)

21

135.Wap to find the nth Fibonacci number.

136.Wap to check whether the number is palindrome or not.

137.wap to print the palindrome numbers between the given range.

138.Wap to find the nth palindrome number.

139.Wap to check whether the number is happy number or not.

140.wap to print the happy numbers between the given range.

141.Wap to find the nth happy number.

142.wap to check whether the string is anagram or not.

143.Wap to check whether the string is pangram or not.

144.Wap to check whether the number is xylem or not. 
                       Ex: 1234→1+4=2+3

145.Wap to check whether the number is spy or not. 
                       Ex: 123→1+2+3=1*2*3 